server/accounts.go

```diff
@@ -1344,26 +1345,85 @@ func (a *Account) AddServiceImportWithClaim(destination *Account, from, to strin
 		return ErrServiceImportAuthorization
 	}
 
-	if a.importFormsCycle(destination, from, to) {
-		return ErrServiceImportFormsCycle
+	// Check if this introduces a cycle before proceeding.
+	if a.serviceImportFormsCycle(destination, from) {
+		return ErrImportFormsCycle
 	}
 
 	_, err := a.addServiceImport(destination, from, to, imClaim)
 	return err
 }
 
-// Detects if we have a cycle.
-func (a *Account) importFormsCycle(destination *Account, from, to string) bool {
-	// Check that what we are importing is not something we also export.
-	if a.serviceExportOverlaps(to) {
-		// So at this point if destination account is also importing from us, that forms a cycle.
-		if destination.serviceImportOverlaps(from) {
+func (a *Account) serviceImportFormsCycle(dest *Account, from string) bool {
+	return dest.checkServiceImportsForCycles(from, map[string]bool{a.Name: true})
+}
+
+func (a *Account) checkServiceImportsForCycles(from string, visited map[string]bool) bool {
+	a.mu.RLock()
+	for _, si := range a.imports.services {
+		if SubjectsCollide(from, si.to) {
+			a.mu.RUnlock()
+			if visited[si.acc.Name] {
+				return true
+			}
+			// Push ourselves and check si.acc
+			visited[a.Name] = true
+			if subjectIsSubsetMatch(si.from, from) {
+				from = si.from
+			}
+			if si.acc.checkServiceImportsForCycles(from, visited) {
```

Comment:
Yes I am ok with recursion here.

Reply:
@derekcollison  Then limit the recursion depth by checking the size of the visited map being smaller than an arbitrary value

Reply:
We need to detect cycles so track down as far as we need, why would we limit? I mean we could and error out but it could be legit (although very unlikely unless someone trying to do something malicious).

Reply:
@derekcollison so that a malicious person can't upload linearly chained accounts and force out to exceed stack space. 

make that pass in a recursion depth vs checking the size of visited.

Reply:
That might be good, but we can just check `len(visited)` I think.